src/test/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepTest.java

```diff
@@ -1161,6 +1161,33 @@ String n(Run<?, ?> b, String label) {
         });
     }
 
+    @Issue("JENKINS-58900")
+    @Test public void nodeDisconnectMissingContextVariableException() {
+        story.then(r -> {
+            DumbSlave agent = r.createOnlineSlave();
+            WorkflowJob p = r.createProject(WorkflowJob.class);
+            p.setDefinition(new CpsFlowDefinition(
+                    "node ('" + agent.getNodeName() + "') {\n" +
+                    "  def isUnix = isUnix()\n" + // Only call `isUnix()` before the agent goes offline to avoid additional log warnings.
+                    "  isUnix ? sh('echo hello') : bat('echo hello')\n" +
+                    "  semaphore('wait')\n" +
+                    "  isUnix ? sh('echo world') : bat('echo hello')\n" +
+                    "}", true));
+            WorkflowRun b = p.scheduleBuild2(0).waitForStart();
+            SemaphoreStep.waitForStart("wait/1", b);
+            agent.toComputer().disconnect(new OfflineCause.UserCause(User.getUnknown(), "going offline"));
```

Comment:
Should we also see `going offline` in the log?

Reply:
Yeah, that would be good, so I added code to print the `OfflineCause` in 8d51aac (copied from `ExecutorStepExecution`) and added an assertion in the test.